File: llm_written/claude-sonnet-4-5/gemm/csharp/test_gemm.cs
using System.Security.Cryptography;
using System.Text;
using System.Diagnostics;

namespace GemmBenchmark;

public class TestGemm
{
    private static Random rng = new Random(42);

    public static double[][] RandomMatrix(int rows, int cols, double min = -1.0, double max = 1.0)
    {
        var result = new double[rows][];
        for (int i = 0; i < rows; i++)
        {
            result[i] = new double[cols];
            for (int j = 0; j < cols; j++)
            {
                result[i][j] = min + (max - min) * rng.NextDouble();
            }
        }
        return result;
    }

    public static double[][] IdentityMatrix(int n)
    {
        var result = Gemm.Zeros(n, n);
        for (int i = 0; i < n; i++)
        {
            result[i][i] = 1.0;
        }
        return result;
    }

    public static string ComputeHash(double[][] matrix)
    {
        using (var sha256 = SHA256.Create())
        {
            var bytes = new List<byte>();
            foreach (var row in matrix)
            {
                foreach (var val in row)
                {
                    bytes.AddRange(BitConverter.GetBytes(val));
                }
            }
            var hash = sha256.ComputeHash(bytes.ToArray());
            return BitConverter.ToString(hash).Replace("-", "").ToLower();
        }
    }

    public static bool MatricesEqual(double[][] A, double[][] B, double tolerance = 0.0)
    {
        if (A.Length != B.Length) return false;
        for (int i = 0; i < A.Length; i++)
        {
            if (A[i].Length != B[i].Length) return false;
            for (int j = 0; j < A[i].Length; j++)
            {
                if (tolerance == 0.0)
                {
                    if (A[i][j] != B[i][j]) return false;
                }
                else
                {
                    if (Math.Abs(A[i][j] - B[i][j]) > tolerance) return false;
                }
            }
        }
        return true;
    }

    public static double[][] DeepCopy(double[][] matrix)
    {
        var result = new double[matrix.Length][];
        for (int i = 0; i < matrix.Length; i++)
        {
            result[i] = new double[matrix[i].Length];
            Array.Copy(matrix[i], result[i], matrix[i].Length);
        }
        return result;
    }

    public class TestCase
    {
        public string Name { get; set; } = "";
        public double[][] A { get; set; } = null!;
        public double[][] B { get; set; } = null!;
        public double Alpha { get; set; } = 1.0;
        public double Beta { get; set; } = 0.0;
        public double[][]? C { get; set; } = null;
    }

    public static List<TestCase> GetTestCases()
    {
        var cases = new List<TestCase>();

        // Edge case: 1x1
        cases.Add(new TestCase
        {
            Name = "1x1",
            A = new double[][] { new double[] { 2.0 } },
            B = new double[][] { new double[] { 3.0 } },
            Alpha = 1.0,
            Beta = 0.0
        });

        // Edge case: 2x2
        cases.Add(new TestCase
        {
            Name = "2x2",
            A = new double[][] { new double[] { 1, 2 }, new double[] { 3, 4 } },
            B = new double[][] { new double[] { 5, 6 }, new double[] { 7, 8 } },
            Alpha = 1.0,
            Beta = 0.0
        });

        // Small: 10x10
        cases.Add(new TestCase
        {
            Name = "10x10",
            A = RandomMatrix(10, 10),
            B = RandomMatrix(10, 10),
            Alpha = 1.0,
            Beta = 0.0
        });

        // Small: non-square 15x20x25
        cases.Add(new TestCase
        {
            Name = "15x20x25",
            A = RandomMatrix(15, 20),
            B = RandomMatrix(20, 25),
            Alpha = 1.0,
            Beta = 0.0
        });

        // Medium: 100x100
        cases.Add(new TestCase
        {
            Name = "100x100",
            A = RandomMatrix(100, 100),
            B = RandomMatrix(100, 100),
            Alpha = 1.0,
            Beta = 0.0
        });

        // Medium: 128x256x128 (tests tiling boundaries)
        cases.Add(new TestCase
        {
            Name = "128x256x128",
            A = RandomMatrix(128, 256),
            B = RandomMatrix(256, 128),
            Alpha = 1.0,
            Beta = 0.0
        });

        // Large: 512x512
        cases.Add(new TestCase
        {
            Name = "512x512",
            A = RandomMatrix(512, 512),
            B = RandomMatrix(512, 512),
            Alpha = 1.0,
            Beta = 0.0
        });

        // Test with alpha and beta
        cases.Add(new TestCase
        {
            Name = "100x100_alpha_beta",
            A = RandomMatrix(100, 100),
            B = RandomMatrix(100, 100),
            Alpha = 2.5,
            Beta = 0.5,
            C = RandomMatrix(100, 100)
        });

        // Test with identity matrix
        cases.Add(new TestCase
        {
            Name = "identity_50x50",
            A = RandomMatrix(50, 50),
            B = IdentityMatrix(50),
            Alpha = 1.0,
            Beta = 0.0
        });

        return cases;
    }

    public static void RunCorrectnessTests(StreamWriter log)
    {
        var cases = GetTestCases();
        int passed = 0;
        int failed = 0;

        log.WriteLine("=== CORRECTNESS TESTS ===");
        log.WriteLine();

        foreach (var testCase in cases)
        {
            try
            {
                // Run sequential
                double[][] seqC = testCase.C != null ? DeepCopy(testCase.C) : null!;
                var seqResult = Gemm.Run(testCase.A, testCase.B, testCase.Alpha, seqC, testCase.Beta);

                // Run parallel
                double[][] parC = testCase.C != null ? DeepCopy(testCase.C) : null!;
                var parResult = GemmParallel.Run(testCase.A, testCase.B, testCase.Alpha, parC, testCase.Beta);

                // Compare
                bool match = MatricesEqual(seqResult, parResult, 0.0);

                if (match)
                {
                    log.WriteLine($"✓ {testCase.Name}: PASS");
                    passed++;
                }
                else
                {
                    log.WriteLine($"✗ {testCase.Name}: FAIL (outputs differ)");
                    failed++;
                }
            }
            catch (Exception ex)
            {
                log.WriteLine($"✗ {testCase.Name}: FAIL (exception: {ex.Message})");
                failed++;
            }
        }

        log.WriteLine();
        log.WriteLine($"Correctness: {passed} passed, {failed} failed");
        log.WriteLine();
    }

    public static void RunDeterminismTests(StreamWriter log)
    {
        log.WriteLine("=== DETERMINISM TESTS ===");
        log.WriteLine();

        var testSizes = new[] { 
            (128, 128, 128),
            (256, 256, 256),
            (512, 512, 512)
        };

        int passed = 0;
        int failed = 0;

        foreach (var (m, k, n) in testSizes)
        {
            var A = RandomMatrix(m, k);
            var B = RandomMatrix(k, n);

            // Run 3 times
            var hashes = new List<string>();
            for (int run = 0; run < 3; run++)
            {
                var result = GemmParallel.Run(A, B);
                var hash = ComputeHash(result);
                hashes.Add(hash);
            }

            // Check all hashes match
            bool allMatch = hashes.All(h => h == hashes[0]);

            if (allMatch)
            {
                log.WriteLine($"✓ {m}x{k}x{n}: DETERMINISTIC");
                log.WriteLine($"  Hash: {hashes[0]}");
                passed++;
            }
            else
            {
                log.WriteLine($"✗ {m}x{k}x{n}: NON-DETERMINISTIC");
                for (int i = 0; i < hashes.Count; i++)
                {
                    log.WriteLine($"  Run {i + 1}: {hashes[i]}");
                }
                failed++;
            }
        }

        log.WriteLine();
        log.WriteLine($"Determinism: {passed} passed, {failed} failed");
        log.WriteLine();
    }

    public static void RunPerformanceTests(StreamWriter log)
    {
        log.WriteLine("=== PERFORMANCE TESTS ===");
        log.WriteLine();

        var testSizes = new[] { 
            (256, 256, 256),
            (512, 512, 512),
            (1024, 1024, 1024)
        };

        foreach (var (m, k, n) in testSizes)
        {
            var A = RandomMatrix(m, k);
            var B = RandomMatrix(k, n);

            // Warmup
            Gemm.Run(A, B);
            GemmParallel.Run(A, B);

            // Sequential timing
            var swSeq = Stopwatch.StartNew();
            var seqResult = Gemm.Run(A, B);
            swSeq.Stop();
            double seqTime = swSeq.Elapsed.TotalSeconds;

            // Parallel timing
            var swPar = Stopwatch.StartNew();
            var parResult = GemmParallel.Run(A, B);
            swPar.Stop();
            double parTime = swPar.Elapsed.TotalSeconds;

            double speedup = seqTime / parTime;
            double efficiency = speedup / Environment.ProcessorCount;

            log.WriteLine($"{m}x{k}x{n}:");
            log.WriteLine($"  Sequential: {seqTime:F3}s");
            log.WriteLine($"  Parallel:   {parTime:F3}s");
            log.WriteLine($"  Speedup:    {speedup:F2}x");
            log.WriteLine($"  Efficiency: {efficiency * 100:F1}% ({Environment.ProcessorCount} cores)");
            log.WriteLine();
        }
    }
}

File: llm_written/claude-sonnet-4-5/gemm/java/TestGemm.java
import java.util.*;
import java.security.MessageDigest;
import java.io.*;

public class TestGemm {

    private static class TestResult {
        String name;
        boolean passed;
        String message;

        TestResult(String name, boolean passed, String message) {
            this.name = name;
            this.passed = passed;
            this.message = message;
        }
    }

    private static String computeHash(double[][] matrix) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            for (double[] row : matrix) {
                for (double val : row) {
                    long bits = Double.doubleToLongBits(val);
                    for (int i = 0; i < 8; i++) {
                        md.update((byte) (bits >> (i * 8)));
                    }
                }
            }
            byte[] digest = md.digest();
            StringBuilder sb = new StringBuilder();
            for (byte b : digest) {
                sb.append(String.format("%02x", b));
            }
            return sb.toString();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static boolean matricesEqual(double[][] A, double[][] B) {
        if (A.length != B.length) return false;
        for (int i = 0; i < A.length; i++) {
            if (A[i].length != B[i].length) return false;
            for (int j = 0; j < A[i].length; j++) {
                if (Double.doubleToLongBits(A[i][j]) != Double.doubleToLongBits(B[i][j])) {
                    return false;
                }
            }
        }
        return true;
    }

    private static double[][] randomMatrix(int rows, int cols, Random rng) {
        double[][] result = new double[rows][cols];
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                result[i][j] = rng.nextDouble() * 10.0 - 5.0;
            }
        }
        return result;
    }

    private static double[][] copyMatrix(double[][] matrix) {
        double[][] result = new double[matrix.length][];
        for (int i = 0; i < matrix.length; i++) {
            result[i] = Arrays.copyOf(matrix[i], matrix[i].length);
        }
        return result;
    }

    private static List<TestResult> runCorrectnessTests() {
        List<TestResult> results = new ArrayList<>();
        Random rng = new Random(42);

        // Test 1: 1x1 matrix
        {
            double[][] A = {{2.0}};
            double[][] B = {{3.0}};
            double[][] seqResult = Gemm.run(A, B);
            double[][] parResult = GemmParallel.run(A, B);
            boolean passed = matricesEqual(seqResult, parResult);
            results.add(new TestResult("1x1 matrix", passed,
                passed ? "PASS" : "FAIL: Results differ"));
        }

        // Test 2: 1xN matrix
        {
            double[][] A = {{1.0, 2.0, 3.0}};
            double[][] B = {{1.0}, {2.0}, {3.0}};
            double[][] seqResult = Gemm.run(A, B);
            double[][] parResult = GemmParallel.run(A, B);
            boolean passed = matricesEqual(seqResult, parResult);
            results.add(new TestResult("1xN matrix", passed,
                passed ? "PASS" : "FAIL: Results differ"));
        }

        // Test 3: Mx1 matrix
        {
            double[][] A = {{1.0}, {2.0}, {3.0}};
            double[][] B = {{1.0, 2.0, 3.0}};
            double[][] seqResult = Gemm.run(A, B);
            double[][] parResult = GemmParallel.run(A, B);
            boolean passed = matricesEqual(seqResult, parResult);
            results.add(new TestResult("Mx1 matrix", passed,
                passed ? "PASS" : "FAIL: Results differ"));
        }

        // Test 4: Small 8x8
        {
            double[][] A = randomMatrix(8, 8, rng);
            double[][] B = randomMatrix(8, 8, rng);
            double[][] seqResult = Gemm.run(A, B);
            double[][] parResult = GemmParallel.run(A, B);
            boolean passed = matricesEqual(seqResult, parResult);
            results.add(new TestResult("8x8 matrix", passed,
                passed ? "PASS" : "FAIL: Results differ"));
        }

        // Test 5: Medium 128x128
        {
            double[][] A = randomMatrix(128, 128, rng);
            double[][] B = randomMatrix(128, 128, rng);
            double[][] seqResult = Gemm.run(A, B);
            double[][] parResult = GemmParallel.run(A, B);
            boolean passed = matricesEqual(seqResult, parResult);
            results.add(new TestResult("128x128 matrix", passed,
                passed ? "PASS" : "FAIL: Results differ"));
        }

        // Test 6: Large 512x512
        {
            double[][] A = randomMatrix(512, 512, rng);
            double[][] B = randomMatrix(512, 512, rng);
            double[][] seqResult = Gemm.run(A, B);
            double[][] parResult = GemmParallel.run(A, B);
            boolean passed = matricesEqual(seqResult, parResult);
            results.add(new TestResult("512x512 matrix", passed,
                passed ? "PASS" : "FAIL: Results differ"));
        }

        // Test 7: Non-square 100x200 * 200x150
        {
            double[][] A = randomMatrix(100, 200, rng);
            double[][] B = randomMatrix(200, 150, rng);
            double[][] seqResult = Gemm.run(A, B);
            double[][] parResult = GemmParallel.run(A, B);
            boolean passed = matricesEqual(seqResult, parResult);
            results.add(new TestResult("100x200 * 200x150", passed,
                passed ? "PASS" : "FAIL: Results differ"));
        }

        // Test 8: With alpha
        {
            double[][] A = randomMatrix(64, 64, rng);
            double[][] B = randomMatrix(64, 64, rng);
            double alpha = 2.5;
            double[][] seqResult = Gemm.run(A, B, alpha);
            double[][] parResult = GemmParallel.run(A, B, alpha);
            boolean passed = matricesEqual(seqResult, parResult);
            results.add(new TestResult("64x64 with alpha=2.5", passed,
                passed ? "PASS" : "FAIL: Results differ"));
        }

        // Test 9: With C and beta
        {
            double[][] A = randomMatrix(64, 64, rng);
            double[][] B = randomMatrix(64, 64, rng);
            double[][] C_seq = randomMatrix(64, 64, rng);
            double[][] C_par = copyMatrix(C_seq);
            double alpha = 1.5;
            double beta = 0.5;
            double[][] seqResult = Gemm.run(A, B, alpha, C_seq, beta);
            double[][] parResult = GemmParallel.run(A, B, alpha, C_par, beta);
            boolean passed = matricesEqual(seqResult, parResult);
            results.add(new TestResult("64x64 with alpha=1.5, beta=0.5", passed,
                passed ? "PASS" : "FAIL: Results differ"));
        }

        // Test 10: alpha = 0
        {
            double[][] A = randomMatrix(64, 64, rng);
            double[][] B = randomMatrix(64, 64, rng);
            double[][] C_seq = randomMatrix(64, 64, rng);
            double[][] C_par = copyMatrix(C_seq);
            double alpha = 0.0;
            double beta = 2.0;
            double[][] seqResult = Gemm.run(A, B, alpha, C_seq, beta);
            double[][] parResult = GemmParallel.run(A, B, alpha, C_par, beta);
            boolean passed = matricesEqual(seqResult, parResult);
            results.add(new TestResult("64x64 with alpha=0", passed,
                passed ? "PASS" : "FAIL: Results differ"));
        }

        return results;
    }

    private static List<TestResult> runDeterminismTests() {
        List<TestResult> results = new ArrayList<>();
        Random rng = new Random(123);

        // Test 1: 512x512 - run twice, compare hashes
        {
            double[][] A = randomMatrix(512, 512, rng);
            double[][] B = randomMatrix(512, 512, rng);
            
            double[][] result1 = GemmParallel.run(A, B);
            String hash1 = computeHash(result1);
            
            double[][] result2 = GemmParallel.run(A, B);
            String hash2 = computeHash(result2);
            
            boolean passed = hash1.equals(hash2);
            results.add(new TestResult("Determinism 512x512 (run 1 vs run 2)", passed,
                passed ? "PASS: hash1=" + hash1.substring(0, 16) + "..., hash2=" + hash2.substring(0, 16) + "..." :
                         "FAIL: hash1=" + hash1 + ", hash2=" + hash2));
        }

        // Test 2: 256x256 - run three times
        {
            double[][] A = randomMatrix(256, 256, rng);
            double[][] B = randomMatrix(256, 256, rng);
            
            double[][] result1 = GemmParallel.run(A, B);
            String hash1 = computeHash(result1);
            
            double[][] result2 = GemmParallel.run(A, B);
            String hash2 = computeHash(result2);
            
            double[][] result3 = GemmParallel.run(A, B);
            String hash3 = computeHash(result3);
            
            boolean passed = hash1.equals(hash2) && hash2.equals(hash3);
            results.add(new TestResult("Determinism 256x256 (3 runs)", passed,
                passed ? "PASS: all hashes match" :
                         "FAIL: hash1=" + hash1 + ", hash2=" + hash2 + ", hash3=" + hash3));
        }

        return results;
    }

    private static void runPerformanceTests() throws IOException {
        Random rng = new Random(456);
        int N = 512;
        int warmupRuns = 2;
        int testRuns = 5;

        double[][] A = randomMatrix(N, N, rng);
        double[][] B = randomMatrix(N, N, rng);

        // Warmup
        for (int i = 0; i < warmupRuns; i++) {
            Gemm.run(A, B);
            GemmParallel.run(A, B);
        }

        // Sequential timing
        long seqTotal = 0;
        for (int i = 0; i < testRuns; i++) {
            long start = System.nanoTime();
            Gemm.run(A, B);
            long end = System.nanoTime();
            seqTotal += (end - start);
        }
        double seqAvg = seqTotal / (double) testRuns / 1e9;

        // Parallel timing
        long parTotal = 0;
        for (int i = 0; i < testRuns; i++) {
            long start = System.nanoTime();
            GemmParallel.run(A, B);
            long end = System.nanoTime();
            parTotal += (end - start);
        }
        double parAvg = parTotal / (double) testRuns / 1e9;

        double speedup = seqAvg / parAvg;
        int cores = Runtime.getRuntime().availableProcessors();
        double efficiency = speedup / cores;

        StringBuilder sb = new StringBuilder();
        sb.append("Performance Test Results\n");
        sb.append("========================\n");
        sb.append("Matrix size: ").append(N).append("x").append(N).append("\n");
        sb.append("Test runs: ").append(testRuns).append("\n");
        sb.append("Sequential time: ").append(String.format("%.4f", seqAvg)).append(" seconds\n");
        sb.append("Parallel time: ").append(String.format("%.4f", parAvg)).append(" seconds\n");
        sb.append("Speedup: ").append(String.format("%.2f", speedup)).append("x\n");
        sb.append("Cores: ").append(cores).append("\n");
        sb.append("Parallel efficiency: ").append(String.format("%.2f", efficiency * 100)).append("%\n");

        System.out.println(sb.toString());

        try (PrintWriter writer = new PrintWriter(new FileWriter("perf.txt"))) {
            writer.print(sb.toString());
        }
    }

    private static void writeSummary(List<TestResult> correctnessResults, List<TestResult> determinismResults) throws IOException {
        StringBuilder sb = new StringBuilder();
        sb.append("GEMM Test Summary\n");
        sb.append("=================\n\n");

        sb.append("Correctness Tests\n");
        sb.append("-----------------\n");
        int correctnessPassed = 0;
        for (TestResult result : correctnessResults) {
            sb.append(result.name).append(": ").append(result.message).append("\n");
            if (result.passed) correctnessPassed++;
        }
        sb.append("Passed: ").append(correctnessPassed).append("/").append(correctnessResults.size()).append("\n\n");

        sb.append("Determinism Tests\n");
        sb.append("-----------------\n");
        int determinismPassed = 0;
        for (TestResult result : determinismResults) {
            sb.append(result.name).append(": ").append(result.message).append("\n");
            if (result.passed) determinismPassed++;
        }
        sb.append("Passed: ").append(determinismPassed).append("/").append(determinismResults.size()).append("\n\n");

        int totalPassed = correctnessPassed + determinismPassed;
        int totalTests = correctnessResults.size() + determinismResults.size();
        sb.append("Overall: ").append(totalPassed).append("/").append(totalTests).append(" tests passed\n");

        System.out.println(sb.toString());

        try (PrintWriter writer = new PrintWriter(new FileWriter("run_summary.txt"))) {
            writer.print(sb.toString());
        }
    }

    public static void main(String[] args) throws IOException {
        if (args.length == 0) {
            System.out.println("Usage: java TestGemm [correctness|determinism|performance|all]");
            System.exit(1);
        }

        String mode = args[0];
        boolean allPassed = true;

        if (mode.equals("correctness") || mode.equals("all")) {
            System.out.println("Running correctness tests...");
            List<TestResult> results = runCorrectnessTests();
            for (TestResult result : results) {
                System.out.println(result.name + ": " + result.message);
                if (!result.passed) allPassed = false;
            }
            if (mode.equals("correctness")) {
                writeSummary(results, new ArrayList<>());
            }
        }

        if (mode.equals("determinism") || mode.equals("all")) {
            System.out.println("\nRunning determinism tests...");
            List<TestResult> results = runDeterminismTests();
            for (TestResult result : results) {
                System.out.println(result.name + ": " + result.message);
                if (!result.passed) allPassed = false;
            }
            if (mode.equals("determinism")) {
                writeSummary(new ArrayList<>(), results);
            }
        }

        if (mode.equals("all")) {
            List<TestResult> correctnessResults = runCorrectnessTests();
            List<TestResult> determinismResults = runDeterminismTests();
            writeSummary(correctnessResults, determinismResults);
            
            for (TestResult r : correctnessResults) {
                if (!r.passed) allPassed = false;
            }
            for (TestResult r : determinismResults) {
                if (!r.passed) allPassed = false;
            }
        }

        if (mode.equals("performance") || mode.equals("all")) {
            System.out.println("\nRunning performance tests...");
            runPerformanceTests();
        }

        System.exit(allPassed ? 0 : 1);
    }
}
